src/hosted-buttons/utils.js

```diff
@@ -0,0 +1,157 @@
+        layout: getButtonVariable(variables, "layout"),
+        shape: getButtonVariable(variables, "shape"),
+        color: getButtonVariable(variables, "color"),
+        label: getButtonVariable(variables, "button_text"),
+      },
+      html: body.html,
+      htmlScript: body.html_script,
+    };
+  });
+};
+
+/**
+ * Attaches form fields (html) to the given selector, and
+ * initializes window.__pp_form_fields (htmlScript).
+ */
+export const renderForm: RenderForm = ({
+  hostedButtonId,
+  html,
+  htmlScript,
+  selector,
+}) => {
+  const elm =
+    typeof selector === "string" ? document.querySelector(selector) : selector;
+  if (elm) {
+    elm.innerHTML = html + htmlScript;
+    const newScriptEl = document.createElement("script");
+    const oldScriptEl = elm.querySelector("script");
+    newScriptEl.innerHTML = oldScriptEl?.innerHTML ?? "";
+    oldScriptEl?.parentNode?.replaceChild(newScriptEl, oldScriptEl);
+  }
+  return {
+    // disable the button, listen for input changes,
+    // and enable the button when the form is valid
+    // using actions.disable() and actions.enable()
+    onInit: window[`__pp_form_fields_${hostedButtonId}`]?.onInit,
+    // render form errors, if present
+    onClick: window[`__pp_form_fields_${hostedButtonId}`]?.onClick,
+  };
+};
+
+export const buildHostedButtonCreateOrder = ({
+  hostedButtonId,
+  merchantId,
+}: GetCallbackProps): CreateOrder => {
+  return (data) => {
+    const userInputs =
+      window[`__pp_form_fields_${hostedButtonId}`]?.getUserInputs?.() || {};
+    return createAccessToken(getClientID()).then((accessToken) => {
+      return request({
+        url: `${apiUrl}/v1/checkout/links/${hostedButtonId}/create-context`,
+        headers: getHeaders(accessToken),
+        method: "POST",
+        body: JSON.stringify({
+          entry_point: entryPoint,
+          funding_source: getFundingSource(data.paymentSource),
+          merchant_id: merchantId,
+          ...userInputs,
+        }),
+      }).then(({ body }) => {
+        return body.context_id;
+      });
+    });
+  };
+};
+
+export const buildHostedButtonOnApprove = ({
+  hostedButtonId,
+  merchantId,
+}: GetCallbackProps): OnApprove => {
+  return (data) => {
+    return createAccessToken(getClientID()).then((accessToken) => {
+      return request({
+        url: `${apiUrl}/v1/checkout/links/${hostedButtonId}/pay`,
+        headers: getHeaders(accessToken),
+        method: "POST",
+        body: JSON.stringify({
+          entry_point: entryPoint,
+          merchant_id: merchantId,
+          context_id: data.orderID,
+        }),
+      }).then((response) => {
+        // The "Debit or Credit Card" button does not open a popup
+        // so we need to open a new popup for buyers who complete
+        // a checkout via "Debit or Credit Card".
+        if (data.paymentSource === "card") {
+          const url = `${baseUrl}/ncp/payment/${hostedButtonId}/${data.orderID}`;
+          popup(url, {
```

Comment:
If running in webview or non popup supported devices this logic would error out. Can smart buttons disable card-form payment flow instead when hosted button id is passed?

Reply:
ah good catch - another option could be redirecting the `window` to this location if `!supportsPopups()` from https://github.com/krakenjs/belter/blob/d0c0868c996aea92d1ad9a6c8662da7a2c200b5b/src/device.js#L229 wdyt?

Reply:
This is also a good option, but effectively this would be duplicate of the logic SPB already has 
 https://github.com/paypal/paypal-smart-payment-buttons/blob/c268be3d0b18ff0ecaa11df646e74c8787ec03ff/src/payment-flows/checkout.js#L653

Also browsers may prevent from opening a popup here, as its not happening inside of a synchronous button onClick event.

Reply:
We chatted offline about this. Keeping it as is in the short term.